File: QID-1223-MVApcabanki/MVApcabanki.r

# clear all variables
rm(list = ls(all = TRUE))
graphics.off()

# load data
x  = read.table("bank2.dat")
n  = nrow(x)
e  = eigen((n - 1) * cov(x)/n)  # calculates eigenvalues and eigenvectors and sorts them by size
e1 = e$values/sum(e$values)

# plot for the relative proportion of variance explained by PCs
dev.new()
plot(e1, ylim = c(0, 0.8), xlab = "Index", ylab = "Variance Explained", main = "Swiss Bank Notes", 
    cex.lab = 1.2, cex.axis = 1.2, cex.main = 1.8)

m    = apply(as.matrix(x), 2, mean)
temp = as.matrix(x - matrix(m, n, ncol(x), byrow = T))
r    = temp %*% e$vectors
r    = cor(cbind(r, x))  # correlation between PCs and variables
r1   = r[7:12, 1:2]      # correlation of the two most important PCs and variables

# plot for the correlation of the original variables with the PCs
dev.new()
ucircle = cbind(cos((0:360)/180 * pi), sin((0:360)/180 * pi))
plot(ucircle, type = "l", lty = "solid", col = "blue", xlab = "First PC", ylab = "Second PC", 
    main = "Swiss Bank Notes", cex.lab = 1.2, cex.axis = 1.2, cex.main = 1.8, lwd = 2)
abline(h = 0, v = 0)
label = c("X1", "X2", "X3", "X4", "X5", "X6")
text(r1, label, cex = 1.2) 

File: QID-1223-MVApcabanki/MVApcabanki.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

x = pd.read_csv("bank2.dat", sep = "\s+", header=None)
n = len(x)

e = np.linalg.eig((n - 1) * np.cov(x.T)/n)
e1 = e[0]/sum(e[0])




fig, ax = plt.subplots(figsize = (10, 10))
ax.scatter(range(1, 7), e1, c = "w", edgecolors = "black")
ax.set_xlabel("Index")
ax.set_ylabel("Variance Explained")
ax.set_title("Swiss Bank Notes")

plt.show()




m = np.mean(x)
temp = np.array(x - np.array([m]*n))
r = np.dot(temp, e[1])
r = pd.concat([pd.DataFrame(r), x], axis = 1).corr()
r1 = r.iloc[6:, :2]

ucircle = pd.DataFrame([np.cos(np.array(range(0, 361))/180 * np.pi), np.sin(np.array(range(0, 361))/180 * np.pi)]).T




fig, ax = plt.subplots(figsize = (10, 10))
ax.plot(ucircle[0], ucircle[1])
ax.hlines(0, -1, 1)
ax.vlines(0, -1, 1)
ax.set_xlabel("First PC")
ax.set_ylabel("Second PC")
for i in range(0, 6):
    plt.text(r1[0][i], r1[1][i], s = "X" + str(i+1))

plt.show()

